### PR TITLE
fix: parse error when only outputs is changed

### LINKED ISSUE #117: parse error
The output of `terraform plan`.

```
Note: Objects have changed outside of Terraform
	
Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # mongodbatlas_cluster.foo has been changed
  ~ resource "mongodbatlas_cluster" "foo" {
        # ...
        # (29 unchanged attributes hidden)



        # (3 unchanged blocks hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────

Changes to Outputs:
  + foo_mongo_uri = "xxx"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: tfplan.binary

To perform exactly these actions, run the following command to apply:
    terraform apply "tfplan.binary"
```

### PR DESCRIPTION
**Pull request by taro-kayo**
Close #117

Related PR: https://github.com/mercari/tfnotify/pull/109

### CODE REVIEW

File: pkg/terraform/parser.go [hunk 1]
```diff
@@ -150,8 +150,8 @@ func (p *PlanParser) Parse(body string) ParseResult { //nolint:cyclop
 		if line == "Terraform will perform the following actions:" { // https://github.com/hashicorp/terraform/blob/332045a4e4b1d256c45f98aac74e31102ace7af7/internal/command/views/plan.go#L252
 			startChangeOutput = i + 1
 		}
-		if startChangeOutput != -1 && endChangeOutput == -1 && strings.HasPrefix(line, "Plan: ") { // https://github.com/hashicorp/terraform/blob/dfc12a6a9e1cff323829026d51873c1b80200757/internal/command/views/plan.go#L306
-			endChangeOutput = i + 1
+		if startChangeOutput != -1 && endChangeOutput == -1 && strings.HasPrefix(line, "-----") { // https://github.com/hashicorp/terraform/blob/1ac7a37d00f3c796f816070847bf02109cb9cab2/internal/command/views/operation.go#L142
```
Comment (taro-kayo) [inline pkg/terraform/parser.go:153]: To include changes to outputs coming after `Plan:`, we have to wait for a horizontal rule which comes after `Changes to Outputs:`.

File: pkg/terraform/parser.go [hunk 2]
```diff
@@ -201,6 +201,10 @@ func (p *PlanParser) Parse(body string) ParseResult { //nolint:cyclop
 
 	changeResult := ""
 	if startChangeOutput != -1 {
+		// if we get here before finding a horizontal rule, output all remaining.
+		if endChangeOutput == -1 {
+			endChangeOutput = len(lines) - 1
+		}
```
Comment (taro-kayo) [inline pkg/terraform/parser.go:207]: If [`TF_IN_AUTOMATION` environment variable](https://developer.hashicorp.com/terraform/tutorials/automation/automate-terraform#controlling-terraform-output-in-automation) is set, no horizontal rules appear after `Terraform will perform the following actions:`. 😢

File: pkg/terraform/parser_test.go [hunk 3]
```diff
@@ -351,6 +455,56 @@ func TestPlanParserParse(t *testing.T) {
 Plan: 1 to add, 0 to change, 0 to destroy.`,
 			},
 		},
+		{
+			name: "plan output changes only pattern 0.12",
+			body: planOnlyOutputChangesSuccessResult0_12,
+			result: ParseResult{
+				Result:             "Plan: 0 to add, 0 to change, 0 to destroy.",
```
Comment (taro-kayo) [inline pkg/terraform/parser_test.go:462]: This is not ideal, but it's very older version... 🙈

**Comment by suzuki-shunsuke** [comment]
Thank you for your contribution!

**Comment by suzuki-shunsuke** [comment]
I'm reviewing this pr yet, but I found an issue.

```console
$ terraform version
Terraform v1.5.7
on darwin_arm64
+ provider registry.terraform.io/hashicorp/github v5.35.0
```

main.tf

```tf
cat main.tf 
import {
  id = "tfaction"
  to = github_repository.tfaction
}

output "foo" {
  value = "foo"
}
```

```console
$ tfcmt plan -- terraform plan

Changes to Outputs:
  + foo = "foo"

You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.
╷
│ Error: Import block target does not exist
│ 
│   on main.tf line 1:
│    1: import {
│ 
│ The target for the given import block does not exist. If you wish to
│ automatically generate config for this resource, use the
│ -generate-config-out option within terraform plan. Otherwise, make sure the
│ target resource exists within your configuration. For example:
│ 
│   terraform plan -generate-config-out=generated.tf
╵
```

AS IS (tfcmt v4.5.0)

image

TO BE

image

At least, the output `Changes to Outputs:` doesn't necessarily mean the success of `terraform plan`.

**Comment by taro-kayo** [comment]
Thanks for testing. 😃 
I'll take a look.

**Comment by taro-kayo** [comment]
OK, I tested.
So, it seems that the behavior in case of error is not this PR problem.
There may be room for improvement regarding the fact that `Click me` is not displayed in the case of `no error, resource no changes, outputs changing`. 😃  

## failed, resource changing, outputs changing

Both v4.5.0 and this PR version have the same result.

> Screenshot 2023-09-09 at 14 32 29

## failed, resource changing, outputs no changes

Same as `failed, resource changing, outputs changing` case above.

## failed, resource no changes, outputs changing

### v4.5.0

> Screenshot 2023-09-09 at 14 37 00

### this PR version

> Screenshot 2023-09-09 at 14 37 42

## no error, resource changing, outputs changing

### v4.5.0

> Screenshot 2023-09-09 at 14 46 30

### this PR version

> Screenshot 2023-09-09 at 14 46 45

## no error, resource changing, outputs no changes

Both v4.5.0 and this PR version have the same result.

> Screenshot 2023-09-09 at 14 42 35

## no error, resource no changes, outputs changing

### v4.5.0

> Screenshot 2023-09-09 at 14 40 45

### this PR version

> Screenshot 2023-09-09 at 14 40 55

**Comment by taro-kayo** [comment]
Fixed the problem I noticed in images in [the comment above](https://github.com/suzuki-shunsuke/tfcmt/pull/896#issuecomment-1712426846). 🎉 

Problem:  `Note: ...` is included in `Change Result`.
fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/e40fa523a3d68eaeb5e554b534fb46f22a06173b

## before

> Screenshot 2023-09-09 at 16 17 37

## after

> Screenshot 2023-09-09 at 16 18 11

---

Problem: `Change Result` is not displayed when only tfstate outputs is changed.
fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/5d98cbaddc2756465bc3535a43b5000d8449ce1f

I copied the test cases from https://github.com/mercari/tfnotify/pull/109, but it may be that Terraform's output specification changed at some point.

## before

> Screenshot 2023-09-09 at 16 19 54


## after

> Screenshot 2023-09-09 at 16 20 14

**Comment by suzuki-shunsuke** [comment]
Thank you for your update.
I think the meaning of the result `Changes to Outputs` isn't clear for users.

image

I'd like to make this message easier to understand.

For example, how about `No resource changes, but Outputs are changed` or `Only Outputs are changed`?

### Plan Result

```
No resource changes, but Outputs are changed
```

### Plan Result

```
Only Outputs are changed
```

**Comment by taro-kayo** [comment]
That's what I thought too, but pretended not to see. 🙈 

Since it is plan, how about changing it to `Only output will be changed.`?
I'll fix it if this is okay.

**Comment by suzuki-shunsuke** [comment]
> Since it is plan, how about changing it to `Only output will be changed.`?

Oh, I see.
Looks good.

Indeed, Terraform uses `will`.

e.g.

> Terraform will perform the following actions:

> null_resource.bar will be created

Could you change to `Only Outputs will be changed.`?

- I'd like to capitalize `Outputs` because this `Outputs` is not a general word but a Terraform specific feature
  - Terraform also capitalize `Outputs`: e.g. `Changes to Outputs`
- I'd like to make `Outputs` plural

**Comment by suzuki-shunsuke** [comment]
Other than that it looks good.

**Comment by taro-kayo** [comment]
`Outputs` looks better. 👍 

fixed: https://github.com/suzuki-shunsuke/tfcmt/pull/896/commits/6fa879f84ecbd3f819343e57b4618db5fbafab78

I'm not too sure about this fix. 😅

**Comment by suzuki-shunsuke** [comment]
Thank you for your great work!

**Comment by suzuki-shunsuke** [comment]
The prerelease version v4.5.1-1 is out.
https://github.com/suzuki-shunsuke/tfcmt/releases/tag/v4.5.1-1

We'll verify this a little, and if there is no problem, we'll release v4.5.1.

**Comment by suzuki-shunsuke** [comment]
v4.5.1 is out 🎉
https://github.com/suzuki-shunsuke/tfcmt/releases/tag/v4.5.1

Thank you for your contribution!

**Pull request merged**